lib/spack/spack/spec.py

```diff
@@ -426,7 +447,56 @@ def __init__(self, spec_like, *dep_like, **kwargs):
         # Spec(a, b) will copy a but just add b as a dep.
         for dep in dep_like:
             spec = dep if isinstance(dep, Spec) else Spec(dep)
-            self._add_dependency(spec)
+            self._add_dependency(spec, (spack.DEP_BUILD, spack.DEP_INSTALL))
+
+
+    def get_dependency(self, name):
+        dep = self._dependencies.get(name)
+        if dep is not None:
+            return dep.spec, dep.deptypes
+        raise InvalidDependencyException(
+            self.name + " does not depend on " + comma_or(name))
+
+
+    def _deptype_norm(self, deptype):
+        if deptype is None:
+            return set(spack.DEP_ALL)
+        # Force deptype to be a tuple so that we can do set intersections.
+        if isinstance(deptype, str):
+            return set((deptype,))
+        return set(deptype)
+
+
+    def _find_deps(self, where, deptype):
+        deptype = self._deptype_norm(deptype)
+
+        return [dep.spec
+                for dep in where.values()
+                if deptype and deptype.intersection(dep.deptypes)]
```

Comment:
Given that there are likely to only be a handful of dep types, I suspect it is way faster to use this construction:

```python
    any(d in deptypes for d in dep.deptypes)
```

Or basically anything that doesn't require building sets.  There are going to be ~3ish compares for each deptype list... which takes way less time than constructing sets for intersection.  I worry about this one b/c we've had issues with lots of object creation in concretization before -- if there end up being a lot of these types of queries and such, it could slow the process down.